service/history/shard/controller_test.go

```diff
@@ -427,6 +427,54 @@ func (s *controllerSuite) TestAcquireShardRenewLookupFailed() {
 	}
 }
 
+func (s *controllerSuite) TestAcquireShard_OwnershipLost() {
+	s.verifyAcquireShardAttempts(&persistence.ShardOwnershipLostError{}, 1, 1)
+}
+
+func (s *controllerSuite) TestAcquireShard_TemporaryErrorIsRetried() {
+	s.verifyAcquireShardAttempts(errors.New("temp error"), 2, 999)
+}
+
+func (s *controllerSuite) verifyAcquireShardAttempts(updateShardErr error, minAttempts int, maxAttempts int) {
+	s.T().Helper()
+	numShards := int32(1)
+	s.config.NumberOfShards = numShards
+	shardID := int32(1)
+	s.mockClusterMetadata.EXPECT().GetCurrentClusterName().Return(cluster.TestCurrentClusterName).AnyTimes()
+	s.mockClusterMetadata.EXPECT().GetAllClusterInfo().Return(cluster.TestSingleDCClusterInfo).AnyTimes()
+	s.mockServiceResolver.EXPECT().AddListener(gomock.Any(), gomock.Any()).Return(nil)
+	s.mockHostInfoProvider.EXPECT().HostInfo().Return(s.hostInfo).AnyTimes()
+	s.mockServiceResolver.EXPECT().Lookup("1").Return(s.hostInfo, nil).AnyTimes()
+	attempts := make(chan struct{}, maxAttempts)
+	s.mockShardManager.EXPECT().GetOrCreateShard(gomock.Any(), getOrCreateShardRequestMatcher(shardID)).Return(
+		&persistence.GetOrCreateShardResponse{
+			ShardInfo: &persistencespb.ShardInfo{
+				ShardId: shardID,
+				Owner:   s.hostInfo.Identity(),
+			},
+		}, nil)
+	s.mockShardManager.EXPECT().UpdateShard(gomock.Any(), gomock.Any()).Return(updateShardErr).
+		AnyTimes().Do(func(_ context.Context, _ *persistence.UpdateShardRequest) {
+		attempts <- struct{}{}
```

Comment:
can you just increment attempts here?

Reply:
That would create a race condition because there would be one routine reading from attempts and another writing to it.